lightning-c-bindings/src/chain/transaction.rs

```diff
@@ -0,0 +1,100 @@
+//! Contains simple structs describing parts of transactions on the chain.
+
+use std::ffi::c_void;
+use bitcoin::hashes::Hash;
+use crate::c_types::*;
+
+
+use lightning::chain::transaction::OutPoint as nativeOutPointImport;
+type nativeOutPoint = nativeOutPointImport;
+
+/// A reference to a transaction output.
+///
+/// Differs from bitcoin::blockdata::transaction::OutPoint as the index is a u16 instead of u32
+/// due to LN's restrictions on index values. Should reduce (possibly) unsafe conversions this way.
+#[must_use]
+#[repr(C)]
+pub struct OutPoint {
+	/// Nearly everyhwere, inner must be non-null, however in places where
+	/// the Rust equivalent takes an Option, it may be set to null to indicate None.
+	pub inner: *mut nativeOutPoint,
+	pub _underlying_ref: bool,
+}
+
+impl Drop for OutPoint {
+	fn drop(&mut self) {
+		if !self._underlying_ref && !self.inner.is_null() {
+			let _ = unsafe { Box::from_raw(self.inner) };
+		}
+	}
+}
+#[no_mangle]
+pub extern "C" fn OutPoint_free(this_ptr: OutPoint) { }
+#[allow(unused)]
+/// Used only if an object of this type is returned as a trait impl by a method
+extern "C" fn OutPoint_free_void(this_ptr: *mut c_void) {
+	unsafe { let _ = Box::from_raw(this_ptr as *mut nativeOutPoint); }
+}
+impl Clone for OutPoint {
+	fn clone(&self) -> Self {
+		Self {
+			inner: Box::into_raw(Box::new(unsafe { &*self.inner }.clone())),
+			_underlying_ref: false,
+		}
+	}
+}
+#[allow(unused)]
+/// Used only if an object of this type is returned as a trait impl by a method
+pub(crate) extern "C" fn OutPoint_clone_void(this_ptr: *const c_void) -> *mut c_void {
+	Box::into_raw(Box::new(unsafe { (*(this_ptr as *mut nativeOutPoint)).clone() })) as *mut c_void
+}
+/// The referenced transaction's txid.
+#[no_mangle]
+pub extern "C" fn OutPoint_get_txid(this_ptr: &OutPoint) -> *const [u8; 32] {
+	let mut inner_val = &mut unsafe { &mut *this_ptr.inner }.txid;
+	(*inner_val).as_inner()
```

Comment:
Why mutable references?

Reply:
This is because all the `inner` pointers are always mut - without it you get the error below, but these mut's allow us to return OpaqueThing { inner: inner_val, ... }.

```
error[E0596]: cannot borrow `*inner_val` as mutable, as it is behind a `&` reference
   --> lightning-c-bindings/src/ln/channelmanager.rs:140:45
    |
139 |     let inner_val = & unsafe { & *this_ptr.inner }.counterparty_features;
    |                     ---------------------------------------------------- help: consider changing this to be a mutable reference: `&mut  unsafe { & *this_ptr.inner }.counterparty_features`
140 |     crate::ln::features::InitFeatures { inner: &mut (*inner_val), _underlying_ref: true }
    |                                                ^^^^^^^^^^^^^^^^^ `inner_val` is a `&` reference, so the data it refers to cannot be borrowed as mutable
```